src/swish/io.ss

```diff
@@ -546,99 +548,109 @@
   (define (make-console-input)
     (binary->utf8 (make-osi-input-port (open-fd-port "stdin-nb" 0 #f))))
 
+  ;; Create a console-input-port that calls the keyboard-interrupt-handler
+  ;; when interrupted via CTRL-C. The Chez Scheme REPL sets the handler to
+  ;; reset and retry the read.
   (define (make-interruptable-console-input)
-    ;; Provide CTRL-C for the repl with a custom binary input port
-    ;; that can be interrupted by CTRL-C to call the
-    ;; keyboard-interrupt-handler and then retry the read just as Chez
-    ;; Scheme does.
+    (define console-process self)
     (define osip (open-fd-port "stdin-nb" 0 #f))
     (define fp 0)
     (define (r! bv start n)
-      (match (with-interrupts-disabled (@r! bv start n))
-        [,count
-         (guard (fixnum? count))
-         (set! fp (+ fp count))
-         count]
-        [interrupt
-         ((keyboard-interrupt-handler))
-         (r! bv start n)]
-        [`(catch ,r) (throw r)]))
+      (let ([count (get-message bv start n)])
+        (set! fp (+ fp count))
+        count))
     (define (gp) fp)
     (define (close) (close-osi-port osip))
+    (define (return v) (void))
+    (define (get-message bv start n)
+      (if (not (eq? self console-process))
+          0 ;; return EOF for console read from another process
+          (let ([r #f] [me self] [request `#(read ,bv ,start ,n)])
+            (send reader request)
+            (on-exit (send reader `#(remove ,request))
+              (let lp ()
+                (with-interrupts-disabled-for-io
+                 (fluid-let ([return (lambda (v) (set! r v) (complete-io me))])
+                   (wait-for-io 'console-input)))
+                (match r
+                  [interrupt (lp)] ;; wait for keyboard-interrupt
+                  [#(io-result ,r) r]
+                  [`(catch ,_) (raise r)]))))))
+    (define reader
+      ;; When CTRL-C or CTRL-BREAK is pressed in Windows, the read returns 0
+      ;; immediately, and later the signal is processed. libuv could return
+      ;; UV_EINTR in this case, but it doesn't. When CTRL-C is pressed in
+      ;; Unix-like systems, the read returns EINTR, and libuv automatically
+      ;; retries.
+      (spawn
+       (lambda ()
+         (define (do-read bv start n)
+           ;; We still assume that each call has the same bv, start, and n,
+           ;; which the Chez Scheme transcoded-port appears to do.
+           (read-osi-port osip bv start n -1))
+         (let lp ([x #f] [timeout 'infinity] [requests '()] [msg #f])
+           (let ([x (or x (receive (after timeout 'timeout) [,x x]))])
+             (match x
+               [#(read ,bv ,start ,n)
+                (lp #f 0 (cons x requests)
```

Comment:
This builds a list of read requests in reverse order, but the list is not processed in reverse order on line 615. Is this a problem? Since there's only one console process that sends read requests one at a time, this list shouldn't go above a single item?

Reply:
Or maybe this is used for nested read calls after CTRL-C? Then the requests should be processed LIFO as the code does now.

Reply:
Yes, I believe the list of `requests` handles nested reads after CTRL-C.

The original code noted that the `transcoded-port` calls `r!` with the same `bv`, `start`, and `n` each time, so we might get away with maintaining a count of pending requests, but we need to get the initial buffer, etc. anyway, hence the list of requests.